### PR TITLE
USM: add AMQP classification on top of TLS

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

This PR adds AMQP classification to USM's TLS classifier. It reuses the classification code used in plaintext, and is expected to behave in the same way, as is verified by the included tests.

I recommend reviewing it commit by commit

### Motivation

Part of the process of matching our TLS classifier with our plaintext classifier.

### Describe how to test/QA your changes

The change is tested by the provided unit tests.

### CODE REVIEW

File: pkg/network/protocols/amqp/testdata/docker-compose.yml [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
       - "RABBITMQ_DEFAULT_USER=${USER:-guest}"
+    volumes:
+      - type: bind
+        source: ${CERTS_PATH}
+        target: /certs
+      - type: bind
+        source: ./${ENCRYPTION_POLICY}.conf
+        target: /etc/rabbitmq/rabbitmq.conf
+
+
```
Comment (guyarb) [inline pkg/network/protocols/amqp/testdata/docker-compose.yml:20]: nit: too many empty lines

File: pkg/network/protocols/amqp/testdata/docker-compose.yml [hunk 2]
```diff
@@ -1,11 +1,20 @@
 version: '3'
-name: amqp
```
Comment (guyarb) [inline pkg/network/protocols/amqp/testdata/docker-compose.yml:2]: why did you remove the name?

File: pkg/network/protocols/amqp/server.go [hunk 3]
```diff
@@ -19,18 +22,67 @@ const (
 	User = "guest"
 	// Pass is the password to use for authentication
 	Pass = "guest"
+
+	// AmqpPlaintext is a flag to indicate that the server should be started with plaintext
+	AmqpPlaintext = false
+	// AmqpTLS is a flag to indicate that the server should be started with TLS
+	AmqpTLS = true
```
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:29]: nit: you're in a package called `amqp` so having `amqp.AmqpPlaintext` is redundant. Use `amqp.Plaintext` instead

File: pkg/network/protocols/amqp/server.go [hunk 4]
```diff
@@ -19,18 +22,67 @@ const (
...
+	cert, _, err := httpUtils.GetCertsPaths()
+	require.NoError(t, err)
+	certsDir := filepath.Dir(cert)
+
+	return []string{
 		"AMQP_ADDR=" + serverAddr,
 		"AMQP_PORT=" + serverPort,
 		"USER=" + User,
 		"PASS=" + Pass,
+		"CERTS_PATH=" + certsDir,
```
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:62]: why is it in the common config?

File: pkg/network/protocols/amqp/server.go
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:1]: nit:
most of the changes can be removed and simplified
```go
var encryptionMode = map[bool]string {TLS: "TLS", Plaintext: "Plaintext"}
var startingRegex = map[bool]*regexp.Regexp {TLS: regexp.MustCompile("...."), ....}

...
func RunServer(t testing.TB, serverAddr, serverPort string, withTLS bool) error {
    ...
    env := commonConfig(t, serverAddr, serverPort)
    env = append(env, fmt.Sprintf("ENCRYPTION_POLICY=%s", encryptionMode[withTLS]))
    return protocolsUtils.RunDockerServer(t, "amqp", dir+"/testdata/docker-compose.yml", env, startingRegex[withTLS], protocolsUtils.DefaultTimeout, 3)
```

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 5]
```diff
@@ -346,6 +346,23 @@ func skipIfHTTPSNotSupported(t *testing.T, _ testContext) {
 	}
 }
 
+func testTLSClassification(t *testing.T, tr *Tracer, clientHost, targetHost, serverHost string) {
+	t.Run("TLS", func(t *testing.T) {
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:358]: I think the wrapper here is redundant, as we have a wrapper per sub-test

File: pkg/network/protocols/amqp/client.go [hunk 6]
```diff
@@ -22,6 +24,12 @@ type Options struct {
 	Username      string
 	Password      string
 	Dialer        *net.Dialer
+
+	// WithTLS indicates whether the connection should be made using TLS
+	WithTLS bool
+	// TlsDialFn is a function that returns a TLS connection as an
+	// io.ReadWriteCloser, suitable to be used with amqp.Open.
+	TLSDialFn func() (io.ReadWriteCloser, error)
```
Comment (guyarb) [inline pkg/network/protocols/amqp/client.go:32]: nit: rename `Tls` to `TLS`

File: pkg/network/protocols/amqp/client.go [hunk 7]
```diff
@@ -173,3 +176,50 @@ func (c *Client) Consume(queue string, numberOfMessages int) ([]string, error) {
...
+
+// newAMQPConnection wraps connection creation from the "amqp" package. It handles
+// the differences in the connection creation process between plaintext & TLS
+// connections. Specifically, for TLS connections, it uses a
+// TransparentUnixProxyServer to handle the TLS part, allowing tests to hook it
+// using USM's GoTLS decoding.
+// Returns a new connection to the AMQP server, on an error if it failed to
+// make one.
+func newAMQPConnection(opts Options) (*amqp.Connection, error) {
+	url := getURL(opts)
```
Comment (guyarb) [inline pkg/network/protocols/amqp/client.go:188]: url is relevant only when TLS is not enabled
thus, please move the declaration near the usage, and I think you should modify `getURL` accordingly, to support only plaintext (and maybe, remove the function, as it wraps a single `fmt.Sprintf` call)

File: pkg/network/ebpf/c/protocols/tls/https.h [hunk 8]
```diff
@@ -47,10 +48,18 @@ static __always_inline void classify_decrypted_payload(protocol_stack_t *stack,
 
     protocol_t proto = PROTOCOL_UNKNOWN;
     classify_protocol_for_dispatcher(&proto, t, buffer, len);
-    if (proto == PROTOCOL_UNKNOWN) {
-        return;
+    if (proto != PROTOCOL_UNKNOWN) {
+        goto update_stack;
+    }
+
+    log_debug("[amqp] starting classification on %p (%lu)", buffer, len);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https.h:55]: Do we need the log here?

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 9]
```diff
@@ -460,3 +614,14 @@ func TestFullMonitorWithTracer(t *testing.T) {
...
+func attachPID(t *testing.T, pid int) {
+	t.Helper()
+	require.NoError(t, usm.GoTLSAttachPID(uint32(pid)))
+	utils.WaitForProgramsToBeTraced(t, "go-tls", pid)
+}
+
+func detachPID(t *testing.T, pid int) {
+	t.Helper()
+	require.NoError(t, usm.GoTLSDetachPID(uint32(pid)))
+}
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:627]: nit: maybe check the program is traced before calling detach (it is not, then silently abort)
and after the detach, check the program is not traced

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 10]
```diff
@@ -460,3 +614,14 @@ func TestFullMonitorWithTracer(t *testing.T) {
 
 	initTracerState(t, tr)
 }
+
+func attachPID(t *testing.T, pid int) {
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:618]: nit: since it is GO-tls only, maybe rename the function

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 11]
```diff
@@ -363,6 +372,151 @@ func testTLSClassification(t *testing.T, tr *Tracer, clientHost, targetHost, ser
 	})
 }
 
+func testTLSAMQPClassification(t *testing.T, tr *Tracer, _, targetHost, serverHost string) {
+	skipFunc := composeSkips(skipIfNotLinux, skipIfUsingNAT, skipIfHTTPSNotSupported)
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:376]: nit: Change `skipIfHTTPSNotSupported` to `skipIfTLSNotSupported`

that can happen in a followup PR

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 12]
```diff
@@ -363,6 +372,151 @@ func testTLSClassification(t *testing.T, tr *Tracer, clientHost, targetHost, ser
...
+				require.NoError(t, err)
+				ctx.extras["client"] = client
+				attachPID(t, cmd.Process.Pid)
+			},
+			postTracerSetup: func(t *testing.T, ctx testContext) {
+				client := ctx.extras["client"].(*amqp.Client)
+				require.NoError(t, client.DeclareQueue("test", client.PublishChannel))
+			},
+			teardown:   amqpTeardown,
+			validation: validateProtocolConnection(&protocols.Stack{Encryption: protocols.TLS}),
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:452]: nit: add a comment in line 429 "declare channel is not supported, thus we expect to classify the protocol as TLS only"

**Review by guyarb — CHANGES_REQUESTED** [review]
overall that's a good PR, the split into commits made it clear, so well done!

I'd add to the PR description and to the code itself, comments explaining why we detach or attach the proxy server in the middle of "preTracer" and "postTracer"

File: pkg/network/protocols/amqp/server.go (hunk 4 above)
Comment (Yumasi) [reply pkg/network/protocols/amqp/server.go:62]: That's because of Docker compose lack of conditionals. Originally it was in the TLS config, but that meant `CERTS_PATH` was not set with the plaintext config, leading to a volume mount with an empty source, which prevented docker compose from starting the container.
I felt this was pretty harmless to do. Another solution would be to use a default value for the variable in docker-compose.yml, but what would that default be?

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 5 above)
Comment (Yumasi) [reply pkg/network/tracer/tracer_usm_linux_test.go:358]: The reason for this "TLS" subtest is to avoid naming conflicts and confusion when looking at the test output:
- currently we already have this test: `TestUSMSuite/.*/TestProtocolClassification/.*/amqp`
- With this subtest, we now have `TestUSMSuite/.*/TestProtocolClassification/.*/TLS/AMQP`
- Without it, we would have `TestUSMSuite/.*/TestProtocolClassification/.*/AMQP` (or, if I had used lowercase for amqp, it would probably show up like this `TestUSMSuite/.*/TestProtocolClassification/.*/amqp#01`).

I think this subtest makes it clear when reading the test report that we are running the tests for AMQP on top of TLS.

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 11 above)
Comment (Yumasi) [reply pkg/network/tracer/tracer_usm_linux_test.go:376]: I am in favor of doing this change in a separate PR, since we have a few places in our code that uses "HTTPS support" instead of "TLS support", not only in the tests.

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 13]
```diff
@@ -615,13 +616,20 @@ func TestFullMonitorWithTracer(t *testing.T) {
...
+func goTLSDetachPID(t *testing.T, pid int) {
 	t.Helper()
+
+	// The program is not traced; nothing to do.
+	if !utils.IsProgramTraced(t, "go-tls", pid) {
+		return
+	}
+
 	require.NoError(t, usm.GoTLSDetachPID(uint32(pid)))
+	require.Falsef(t, utils.IsProgramTraced(t, "go-tls", pid), "process %v is still traced by Go-TLS after detaching", pid)
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:634]: let's use retry mechanism here to ensure the test is not flaky

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/tracer/tracer_usm_linux_test.go:358]: nit: let's use a single pattern - either AMQP or amqp

**Review by guyarb — APPROVED** [review]

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 14]
```diff
@@ -99,6 +105,12 @@ func classificationSupported(config *config.Config) bool {
 	return kprobe.ClassificationSupported(config)
 }
 
+func skipIfGoTLSNotSupported(t *testing.T, ctx testContext) {
+	if !gotlstestutil.GoTLSSupported(t, ctx.extras["config"].(*config.Config)) {
+		t.Skip("GoTLS is not supported")
+	}
+}
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:112]: no need for `ctx.extras["config"].(*config.Config)`

The configuration we need in `GoTLSSupported` needs `EnableRuntimeCompiler` and `EnableCORE`. you can get both from calling `config.New()`

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 15]
```diff
@@ -1654,44 +1717,58 @@ func testAMQPProtocolClassification(t *testing.T, tr *Tracer, clientHost, target
...
+		utils.WaitForProgramsToBeTraced(t, "go-tls", pid)
+
+		preTracerSetupWrapper = func(f func(t *testing.T, ctx testContext)) func(*testing.T, testContext) {
+			return func(t *testing.T, ctx testContext) {
+				goTLSDetachPID(t, pid)
+				f(t, ctx)
+				goTLSAttachPID(t, pid)
+			}
+		}
+	}
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1738]: 1. No need to rerun the proxy server every time, you all need to attach or detach in each test
2. Why keeping a large difference between plaintext and TLS tests? you can use the proxy for all scenarios, and just enable/disable TLS

Comment (guyarb) [reply pkg/network/tracer/tracer_usm_linux_test.go:1738]: we need to detach after the postTracer

File: pkg/network/protocols/amqp/client.go [hunk 16]
```diff
@@ -173,3 +176,39 @@ func (c *Client) Consume(queue string, numberOfMessages int) ([]string, error) {
...
+
+	if opts.WithTLS {
+		return amqp.DialTLS(url, &tls.Config{InsecureSkipVerify: true})
+	}
+
+	dialOptions := amqp.Config{}
+	if opts.Dialer != nil {
+		dialOptions.Dial = opts.Dialer.Dial
+	}
```
Comment (guyarb) [inline pkg/network/protocols/amqp/client.go:197]: is it still needed?

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 17]
```diff
@@ -153,6 +154,7 @@ func (s *USMSuite) TestProtocolClassification() {
 	}
 
 	cfg.EnableNativeTLSMonitoring = true
+	cfg.EnableGoTLSSupport = gotlstestutil.GoTLSSupported(t, cfg)
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:157]: redundant, appears in line 160

**Review by guyarb — APPROVED** [review]
